### PR TITLE
FIX Make  _validate_and_reformat_input less strict

### LINKED ISSUE #898: Bug in mitigation using Fairlearn on Azure AutoML models
#### Describe the bug

`GridSearch` is being too restrictive when validating its inputs, rejecting data which it should be able to handle

#### Steps/Code to Reproduce

```python
from sklearn.model_selection import train_test_split
from fairlearn.reductions import GridSearch
from fairlearn.reductions import DemographicParity, ErrorRate
from fairlearn.metrics import MetricFrame, selection_rate, count
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn import metrics as skm
import pandas as pd


from sklearn.datasets import fetch_openml
data = fetch_openml(data_id=1590, as_frame=True)
X_raw = data.data
Y = (data.target == '>50K') * 1


A = X_raw["sex"]
X = X_raw.drop(labels=['sex'], axis=1)

le = LabelEncoder()
Y = le.fit_transform(Y)

X_train, X_test, Y_train, Y_test, A_train, A_test = train_test_split(X,
                                                                     Y,
                                                                     A,
                                                                     test_size=0.2,
                                                                     random_state=0,
                                                                     stratify=Y)

# Work around indexing bug
X_train = X_train.reset_index(drop=True)
A_train = A_train.reset_index(drop=True)
X_test = X_test.reset_index(drop=True)
A_test = A_test.reset_index(drop=True)



# The following shows that LGBMClassifier can handle this dataset
# without the scaling and one-hot encoding used in other examples
from lightgbm import LGBMClassifier

unmitigated_predictor = LGBMClassifier()
unmitigated_predictor.fit(X_train, Y_train)

metric_frame = MetricFrame(metrics={"accuracy": skm.accuracy_score,
                                    "selection_rate": selection_rate,
                                    "count": count},
                           sensitive_features=A_test,
                           y_true=Y_test,
                           y_pred=unmitigated_predictor.predict(X_test))
print(metric_frame.overall)
print(metric_frame.by_group)

# However a GridSearch fails:
sweep = GridSearch(LGBMClassifier(),
                   constraints=DemographicParity(),
                   grid_size=71)

sweep.fit(X_train, Y_train,
          sensitive_features=A_train)

```

#### Expected Results

That the `sweep.fit()` call works

#### Actual Results

```
c:\users\riedgar\appdata\local\continuum\miniconda3\envs\nb-automl-fairlearn\lib\site-packages\sklearn\utils\validation.py in check_array(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, estimator)
    709             try:
--> 710                 array = array.astype(np.float64)
    711             except ValueError as e:

ValueError: could not convert string to float: 'Private'

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
<ipython-input-9-cb592bed5a23> in <module>
      1 sweep.fit(X_train, Y_train,
----> 2           sensitive_features=A_train)
      3 
      4 predictors = sweep.predictors_

c:\users\riedgar\appdata\local\continuum\miniconda3\envs\nb-automl-fairlearn\lib\site-packages\fairlearn\reductions\_grid_search\grid_search.py in fit(self, X, y, **kwargs)
    135         # Prep the parity constraints and objective
    136         logger.debug("Preparing constraints and objective")
--> 137         self.constraints.load_data(X, y, **kwargs)
    138         objective = self.constraints.default_objective()
    139         objective.load_data(X, y, **kwargs)

c:\users\riedgar\appdata\local\continuum\miniconda3\envs\nb-automl-fairlearn\lib\site-packages\fairlearn\reductions\_moments\utility_parity.py in load_data(self, X, y, sensitive_features, control_features)
    309                                          enforce_binary_labels=True,
    310                                          sensitive_features=sensitive_features,
--> 311                                          control_features=control_features)
    312 
    313         base_event = pd.Series(data=_ALL, index=y_train.index)

c:\users\riedgar\appdata\local\continuum\miniconda3\envs\nb-automl-fairlearn\lib\site-packages\fairlearn\utils\_input_validation.py in _validate_and_reformat_input(X, y, expect_y, enforce_binary_labels, **kwargs)
     70             y = y.to_numpy().reshape(-1)
     71 
---> 72         X, y = check_X_y(X, y)
     73         y = check_array(y, ensure_2d=False, dtype='numeric')
     74         if enforce_binary_labels and not set(np.unique(y)).issubset(set([0, 1])):
```

Obviously our call to `check_X_y()` is too restrictive.

### PR DESCRIPTION
**Pull request by riedgar-ms**
The call to `check_X_y()` in `_validate_and_reformat_input()` was unduly harsh, and prevented `GridSearch` from running with a model which didn't mind non-numeric columns.

Closes #898 

Signed-off-by: Richard Edgar <riedgar@microsoft.com>

### CODE REVIEW

**Comment by romanlutz** [comment]
Do you want to add a test case? Otherwise someone might come along in the future and undo this without realizing what it was for.

**Comment by riedgar-ms** [comment]
> Do you want to add a test case? Otherwise someone might come along in the future and undo this without realizing what it was for.

I shall see what I can concoct. It won't be much of a test, since this is about loosening restrictions.

**Comment by romanlutz** [comment]
> > Do you want to add a test case? Otherwise someone might come along in the future and undo this without realizing what it was for.
> 
> I shall see what I can concoct. It won't be much of a test, since this is about loosening restrictions.

That's what I mean 🙂 There should be a test for whatever new scenarios you're enabling, so that it would fail if you remove the changes you're adding here.

**Comment by riedgar-ms** [comment]
Test added @romanlutz ..... not one of the best, but does at least make sure that we don't enforce X to be all numeric.

**Review by romanlutz — APPROVED** [review]

**Comment by adrinjalali** [comment]
what do we do with this?

> No hosted parallelism has been purchased or granted. To request a free parallelism grant, please fill out the following form https://aka.ms/azpipelines-parallelism-request

File: test/unit/utils/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+# Licensed under the MIT License.
+
+import numpy as np
+import pandas as pd
+
+import fairlearn.utils._input_validation as iv
+
+
+class TestValidateAndReformatInput:
+    def test_smoke(self):
```
Comment (adrinjalali) [inline test/unit/utils/test_utils.py:11]: it'd be nice to have a comment like `regression test for #906`

**Review by adrinjalali — COMMENTED** [review]
Otherwise LGTM.

File: test/unit/utils/test_utils.py (hunk 1 above)
Comment (riedgar-ms) [reply test/unit/utils/test_utils.py:11]: Done

**Comment by romanlutz** [comment]
> what do we do with this?
> 
> > No hosted parallelism has been purchased or granted. To request a free parallelism grant, please fill out the following form https://aka.ms/azpipelines-parallelism-request

@adrinjalali this is due to https://github.com/fairlearn/fairlearn/issues/757#issuecomment-881224903 and not a problem. It's just duplicating the build that's run in the other Azure org. We can override for the merge if everything else passes.

File: fairlearn/utils/_input_validation.py [hunk 2]
```diff
@@ -69,7 +69,7 @@ def _validate_and_reformat_input(X, y=None, expect_y=True, enforce_binary_labels
         elif isinstance(y, pd.DataFrame) and y.shape[1] == 1:
             y = y.to_numpy().reshape(-1)
 
-        X, y = check_X_y(X, y)
+        X, y = check_X_y(X, y, dtype=None, force_all_finite=False)
         y = check_array(y, ensure_2d=False, dtype='numeric')
```
Comment (adrinjalali) [inline fairlearn/utils/_input_validation.py:73]: does `y` really need to be `numeric`? Or is it also something the underlying estimator should check? I think for a classification `y` could be categorical?

Comment (riedgar-ms) [reply fairlearn/utils/_input_validation.py:73]: I believe that the way the reweighting currently works, it has to be 0 or 1. I agree that it ought to be possible to take any two categorical values, but that would be a bigger change.

**Review by adrinjalali — APPROVED** [review]

**Pull request merged**